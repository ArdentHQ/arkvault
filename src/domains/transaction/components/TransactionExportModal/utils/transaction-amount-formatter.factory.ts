import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO, Helpers } from "@ardenthq/sdk-profiles";

const formatAmount = (amount: number, transaction: DTO.ExtendedConfirmedTransactionData) =>
	Helpers.Currency.format(amount, transaction.wallet().exchangeCurrency(), {
		withTicker: false,
	});

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isReceived()) {
		let totalReceived = BigNumber.make(transaction.amount());

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived.minus(recipient.amount);
			}
		}

		return formatAmount(totalReceived.toNumber(), transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return multiPaymentAmount(transaction);
	}

	return formatAmount(transaction.amount(), transaction);
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	if (transaction.isMultiPayment()) {
		return BigNumber.make(multiPaymentAmount(transaction)).plus(transaction.fee());
	}

	return formatAmount(transaction.total(), transaction);
};

export const AmountFormatter = (transaction: DTO.ExtendedConfirmedTransactionData) => ({
	amount: () => transactionAmount(transaction),
	fee: () => formatAmount(transaction.fee(), transaction),
	total: () => transactionTotal(transaction),
});
