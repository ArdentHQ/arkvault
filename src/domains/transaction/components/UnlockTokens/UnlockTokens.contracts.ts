import { Services } from "@ardenthq/sdk";
import { Contracts } from "@ardenthq/sdk-profiles";

const POLLING_INTERVAL = 1000 * 60; // 1 min

enum Step {
	SelectStep,
	ReviewStep,
	AuthenticationStep,
	SummaryStep,
	ErrorStep,
}

interface UnlockableBalance extends Services.UnlockableBalance {
	id: string;
}

interface UnlockTokensFormState {
	fee: number;
	amount: number;
	selectedObjects: UnlockableBalance[];

	// authentication fields
	mnemonic: string | undefined;
	secondMnemonic: string | undefined;
	encryptionPassword: string | undefined;
	wif: string | undefined;
	privateKey: string | undefined;
	secret: string | undefined;
}

type UseUnlockableBalancesHook = (wallet: Contracts.IReadWriteWallet) => {
	items: UnlockableBalance[];
	loading: boolean;
	isFirstLoad: boolean;
};

export type { UnlockableBalance, UnlockTokensFormState, UseUnlockableBalancesHook };

export interface UnlockTokensModalProperties {
	profile: Contracts.IProfile;
	wallet: Contracts.IReadWriteWallet;
	onClose: () => void;
}

export { POLLING_INTERVAL, Step };
