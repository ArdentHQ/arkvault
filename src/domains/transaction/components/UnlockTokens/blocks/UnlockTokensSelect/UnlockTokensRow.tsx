import React, { FC } from "react";
import { useTranslation } from "react-i18next";

import { UnlockTokensRowSkeleton } from "./UnlockTokensRowSkeleton";
import { UnlockTokensRowProperties } from "./UnlockTokensSelect.contracts";
import { Amount } from "@/app/components/Amount";
import { Checkbox } from "@/app/components/Checkbox";
import { Icon } from "@/app/components/Icon";
import { TableCell, TableRow } from "@/app/components/Table";

export const UnlockTokensRow: FC<UnlockTokensRowProperties> = ({ loading, item, ticker, onToggle, checked }) => {
	const { t } = useTranslation();

	if (loading) {
		return <UnlockTokensRowSkeleton />;
	}

	const date = item.timestamp.format("DD MMM YYYY");
	const time = item.timestamp.format("HH:mm:ss");
	const relativeTime = item.timestamp.fromNow();

	return (
		<TableRow>
			<TableCell variant="start" isCompact>
				<Amount
					className="font-bold text-theme-secondary-900 dark:text-theme-secondary-200"
					value={item.amount.toHuman()}
					ticker={ticker}
				/>
			</TableCell>

			<TableCell isCompact>
				<time className="text-theme-secondary-700" dateTime={item.timestamp.toDate().toUTCString()}>
					{date} {t("COMMON.AT")} {time} ({relativeTime})
				</time>
			</TableCell>

			<TableCell variant="end" isCompact innerClassName="justify-end">
				{item.isReady ? (
					<div className="flex items-center">
						<div className="flex items-center">
							<span className="text-theme-secondary-700" data-testid="UnlockableBalanceRow__status">
								{t("TRANSACTION.UNLOCK_TOKENS.UNLOCKABLE")}
							</span>
							<Icon name="LockOpen" size="lg" className="mr-3 ml-2 text-theme-primary-600" />
						</div>
						<Checkbox checked={checked} onChange={onToggle} />
					</div>
				) : (
					<div className="flex items-center">
						<div className="flex items-center">
							<span className="text-theme-secondary-700" data-testid="UnlockableBalanceRow__status">
								{t("TRANSACTION.UNLOCK_TOKENS.LOCKED")}
							</span>
							<Icon name="Lock" size="lg" className="mr-3 ml-2 text-theme-secondary-700" />
						</div>
						<Checkbox checked={false} disabled />
					</div>
				)}
			</TableCell>
		</TableRow>
	);
};
