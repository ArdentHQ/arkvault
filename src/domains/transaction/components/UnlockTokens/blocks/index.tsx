export { UnlockTokensAuthentication } from "./UnlockTokensAuthentication";
export { UnlockTokensReview } from "./UnlockTokensReview";
export { UnlockTokensSelect } from "./UnlockTokensSelect";
export { UnlockTokensSummary } from "./UnlockTokensSummary";
