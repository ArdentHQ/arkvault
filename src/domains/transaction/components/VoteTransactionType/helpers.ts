import { DTO } from "@ardenthq/sdk";
import { DTO as ProfilesDTO } from "@ardenthq/sdk-profiles";

/**
 * @TODO: This can be handled in SDK to retrieve voting and unvoting public keys
 * through a common method/format, whether it's a signed or confirmed transaction.
 * Currently votes() unvotes() methods don't exist signed transaction data
 *  and `data` is a function in signed transaction and an object in confirmed transaction
 */
export const extractVotingData = ({ transaction }: { transaction: DTO.RawTransactionData }) => {
	if (transaction.isConfirmed()) {
		return {
			unvotes: transaction.unvotes(),
			votes: transaction.votes(),
		};
	}

	return signedTransactionVotingData({ transaction });
};

const signedTransactionVotingData = ({ transaction }: { transaction: ProfilesDTO.ExtendedSignedTransactionData }) => {
	const data = transaction.data().data();
	const votes: string[] = [];
	const unvotes: string[] = [];

	// array of publicKeys
	const rawVotes: string[] = data?.asset?.votes ?? [];

	for (const publicKey of rawVotes) {
		const destination = publicKey.startsWith("-") ? unvotes : votes;
		destination.push(publicKey.replace(/^[+-]+/, ""));
	}

	return {
		unvotes,
		votes,
	};
};
