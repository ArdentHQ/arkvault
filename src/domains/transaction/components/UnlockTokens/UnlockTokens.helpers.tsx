import { sortByDesc } from "@payvo/sdk-helpers";
import React, { useMemo, useRef, useState } from "react";
import { useTranslation } from "react-i18next";
import { Column } from "react-table";

import { UnlockTokensFetchError } from "./blocks/UnlockTokensFetchError";
import { POLLING_INTERVAL, UnlockableBalance, UseUnlockableBalancesHook } from "./UnlockTokens.contracts";
import { Checkbox } from "@/app/components/Checkbox";
import { Tooltip } from "@/app/components/Tooltip";
import { useScheduler } from "@/app/hooks/use-scheduler";
import { toasts } from "@/app/services";

const useUnlockableBalances: UseUnlockableBalancesHook = (wallet) => {
	const loadCount = useRef(0);

	const { start, stop } = useScheduler({
		autostart: true,
		handler: () => fetch(),
		timeout: POLLING_INTERVAL,
	});

	const [items, setItems] = useState<UnlockableBalance[]>([]);
	const [loading, setLoading] = useState<boolean>(false);

	const fetch = async () => {
		setLoading(true);

		try {
			const response = await wallet.coin().client().unlockableBalances(wallet.address());

			setItems(
				sortByDesc(response.objects, (value) => value.timestamp.toUNIX()).map((value, index) => ({
					...value,
					id: `${index}`,
				})),
			);

			loadCount.current++;
		} catch {
			stop();

			setItems([]);

			toasts.warning(
				<UnlockTokensFetchError
					onRetry={
						/* istanbul ignore next */ () => {
							void toasts.dismiss();

							start();
						}
					}
				/>,
			);
		} finally {
			setLoading(false);
		}
	};

	return {
		isFirstLoad: loadCount.current === 1,
		items,
		loading,
	};
};

const useUnlockTokensSelectTableColumns = (
	isSelectDisabled: boolean,
	isAllSelected: boolean,
	onToggleAll: () => void,
) => {
	const { t } = useTranslation();

	return useMemo<Column<UnlockableBalance>[]>(
		() => [
			{
				Header: t("COMMON.AMOUNT"),
				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
				accessor: (unlockableBalance) => unlockableBalance.amount?.toNumber(),
				id: "amount",
			},
			{
				Header: t("COMMON.TIME"),
				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
				accessor: (unlockableBalance) => unlockableBalance.timestamp?.toUNIX(),
				id: "time",
			},
			{
				Header: (
					<>
						<span className="mr-3">{t("COMMON.STATUS")}</span>
						<Tooltip content={isAllSelected ? t("COMMON.UNSELECT_ALL") : t("COMMON.SELECT_ALL")}>
							<Checkbox disabled={isSelectDisabled} checked={isAllSelected} onChange={onToggleAll} />
						</Tooltip>
					</>
				),
				className: "justify-end float-right",
				id: "status",
			},
		],
		[t, isSelectDisabled, isAllSelected, onToggleAll],
	);
};

export { useUnlockableBalances, useUnlockTokensSelectTableColumns };
