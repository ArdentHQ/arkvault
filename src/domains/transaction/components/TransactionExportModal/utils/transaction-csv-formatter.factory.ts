import { BigNumber } from "@ardenthq/sdk-helpers";
import { DTO } from "@ardenthq/sdk-profiles";
import { CURRENCIES } from "@ardenthq/sdk-intl";
import { buildTranslations } from "@/app/i18n/helpers";

const recipient = (transaction: DTO.ExtendedConfirmedTransactionData) => {
	const { COMMON } = buildTranslations();

	if (transaction.isMultiPayment()) {
		return COMMON.MULTIPLE;
	}

	if (transaction.isTransfer()) {
		return transaction.recipient();
	}

	if (transaction.isVote() || transaction.isUnvote()) {
		return `${COMMON.VOTE} ${COMMON.TRANSACTION}`;
	}

	return COMMON.OTHER;
};

const multiPaymentAmount = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.sender() !== transaction.wallet().address()) {
		let totalReceived = BigNumber.ZERO;

		for (const recipient of transaction.recipients()) {
			if (recipient.address === transaction.wallet().address()) {
				totalReceived = totalReceived.plus(recipient.amount);
			}
		}

		return totalReceived.toNumber();
	}

	return transaction.amount();
};

const transactionAmount = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	const amount = transaction.isMultiPayment() ? multiPaymentAmount(transaction) : transaction.amount();

	if (transaction.isSent()) {
		return BigNumber.make(amount).times(-1).toNumber();
	}

	return amount;
};

const transactionFee = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.isSent()) {
		return BigNumber.make(transaction.fee()).times(-1).toNumber();
	}

	return 0;
};

const transactionTotal = (transaction: DTO.ExtendedConfirmedTransactionData): number => {
	if (transaction.isSent()) {
		return BigNumber.make(transaction.total()).times(-1).toNumber();
	}

	return transactionAmount(transaction);
};

const converted = (value: number, rate: number) => BigNumber.make(value).times(rate).toNumber();

const truncate = (value: number, currency: string) => {
	const decimals = CURRENCIES[currency]?.decimals ?? 8;

	return Math.trunc(value * 10 ** decimals) / 10 ** decimals;
};

export const CsvFormatter = (transaction: DTO.ExtendedConfirmedTransactionData, timeFormat: string) => {
	const amount = transactionAmount(transaction);
	const fee = transactionFee(transaction);
	const total = transactionTotal(transaction);

	const currency = transaction.wallet().currency();
	const exchangeCurrency = transaction.wallet().exchangeCurrency();

	const rate =
		transaction.total() === 0
			? 0
			: truncate(
					BigNumber.make(transaction.convertedTotal()).divide(transaction.total()).toNumber(),
					exchangeCurrency,
			  );

	return {
		amount: () => truncate(amount, currency),
		convertedAmount: () => truncate(converted(amount, rate), exchangeCurrency),
		convertedFee: () => (fee === 0 ? 0 : truncate(converted(fee, rate), exchangeCurrency)),
		convertedTotal: () => truncate(converted(total, rate), exchangeCurrency),
		datetime: () => transaction.timestamp()?.format(`DD.MM.YYYY ${timeFormat}`),
		fee: () => fee,
		rate: () => rate,
		recipient: () => recipient(transaction),
		sender: () => transaction.sender(),
		timestamp: () => transaction.timestamp()?.toUNIX(),
		total: () => truncate(total, currency),
	};
};
