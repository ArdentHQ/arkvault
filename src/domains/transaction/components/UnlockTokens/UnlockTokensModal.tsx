import { Services } from "@ardenthq/sdk";
import { DTO } from "@ardenthq/sdk-profiles";
import React, { useEffect, useRef, useState } from "react";
import { SubmitHandler, useForm } from "react-hook-form";

import { UnlockTokensAuthentication, UnlockTokensReview, UnlockTokensSelect, UnlockTokensSummary } from "./blocks";
import { Step, UnlockTokensFormState, UnlockTokensModalProperties } from "./UnlockTokens.contracts";
import { useUnlockableBalances } from "./UnlockTokens.helpers";
import { Form } from "@/app/components/Form";
import { Modal } from "@/app/components/Modal";
import { useEnvironmentContext, useLedgerContext } from "@/app/contexts";
import { ErrorStep } from "@/domains/transaction/components/ErrorStep";
import { useTransactionBuilder } from "@/domains/transaction/hooks";
import { handleBroadcastError } from "@/domains/transaction/utils";

export const UnlockTokensModal: React.VFC<UnlockTokensModalProperties> = ({ profile, wallet, onClose }) => {
	const [step, setStep] = useState<Step>(Step.SelectStep);

	const [transaction, setTransaction] = useState<DTO.ExtendedSignedTransactionData | undefined>(undefined);
	const [error, setError] = useState<Error | undefined>();

	const { persist } = useEnvironmentContext();

	const abortReference = useRef<AbortController | undefined>();
	const transactionBuilder = useTransactionBuilder();

	const { hasDeviceAvailable, isConnected, connect } = useLedgerContext();

	const { items, loading, isFirstLoad } = useUnlockableBalances(wallet);

	const form = useForm<UnlockTokensFormState>({
		defaultValues: {
			amount: 0,
			fee: 0,
			selectedObjects: [],
		},
		mode: "onChange",
	});

	const { register, handleSubmit, formState } = form;

	useEffect(() => {
		register("amount");
		register("fee");
		register("selectedObjects");
	}, [register]);

	const submit: SubmitHandler<UnlockTokensFormState> = async ({ selectedObjects, ...authenticationData }) => {
		try {
			const signatory = await wallet.signatoryFactory().make(authenticationData);

			const input: Services.UnlockTokenInput = {
				data: { objects: selectedObjects },
				signatory,
			};

			/* istanbul ignore next: Ledger signing of this TX type is not allowed yet for the coins we support */
			if (wallet.isLedger()) {
				await connect(profile, wallet.coinId(), wallet.networkId());
			}

			abortReference.current = new AbortController();

			const { uuid, transaction } = await transactionBuilder.build("unlockToken", input, wallet, {
				abortSignal: abortReference.current.signal,
			});

			const response = await wallet.transaction().broadcast(uuid);

			handleBroadcastError(response);

			await wallet.transaction().sync();
			await persist();

			setTransaction(transaction);
			setStep(Step.SummaryStep);
		} catch (error) {
			setError(error)
			setStep(Step.ErrorStep);
		}
	};

	return (
		<Modal isOpen size="3xl" title={undefined} onClose={onClose}>
			<Form onSubmit={submit} context={form} data-testid="UnlockTokensModal">
				{step === Step.SelectStep && (
					<UnlockTokensSelect
						items={items}
						loading={loading}
						isFirstLoad={isFirstLoad}
						profile={profile}
						wallet={wallet}
						onClose={onClose}
						onUnlock={() => setStep(Step.ReviewStep)}
					/>
				)}

				{step === Step.ReviewStep && (
					<UnlockTokensReview
						wallet={wallet}
						onBack={() => setStep(Step.SelectStep)}
						onConfirm={() => setStep(Step.AuthenticationStep)}
					/>
				)}

				{step === Step.AuthenticationStep && (
					<UnlockTokensAuthentication
						ledgerIsAwaitingApp={!isConnected}
						ledgerIsAwaitingDevice={!hasDeviceAvailable}
						wallet={wallet}
						onBack={() => {
							abortReference.current?.abort();
							setStep(Step.ReviewStep);
						}}
					/>
				)}

				{step === Step.SummaryStep && !!transaction && <UnlockTokensSummary transaction={transaction} />}

				{step === Step.ErrorStep && (
					<ErrorStep
						error={error}
						onRepeat={/* istanbul ignore next */ () => handleSubmit(submit)}
						isRepeatDisabled={formState.isSubmitting}
					/>
				)}
			</Form>
		</Modal>
	);
};
