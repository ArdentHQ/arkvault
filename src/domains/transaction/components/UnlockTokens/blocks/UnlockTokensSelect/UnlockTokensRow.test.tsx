import { BigNumber } from "@payvo/sdk-helpers";
import { DateTime } from "@payvo/sdk-intl";
import userEvent from "@testing-library/user-event";
import React from "react";

import { UnlockTokensRow } from "./UnlockTokensRow";
import { buildTranslations } from "@/app/i18n/helpers";
import { UnlockableBalance } from "@/domains/transaction/components/UnlockTokens/UnlockTokens.contracts";
import { render, screen } from "@/utils/testing-library";

const translations = buildTranslations();

describe("UnlockTokensRow", () => {
	const item: UnlockableBalance = {
		address: "lskbps7ge5n9y7f8nk4222c77zkqcntrj7jyhmkwp",
		amount: BigNumber.make(10),
		height: "123456",
		id: "1",
		isReady: true,
		timestamp: DateTime.make("2020-01-01"),
	};

	it("should render loading", () => {
		render(
			<table>
				<tbody>
					<UnlockTokensRow loading item={{}} ticker="LSK" checked={false} onToggle={jest.fn()} />
				</tbody>
			</table>,
		);

		expect(screen.queryByTestId("Amount")).toBeNull();
	});

	it.each([true, false])("should render with isReady = %s", (isReady) => {
		const onToggle = jest.fn();

		const { asFragment } = render(
			<table>
				<tbody>
					<UnlockTokensRow
						loading={false}
						item={{ ...item, isReady }}
						ticker="LSK"
						checked
						onToggle={onToggle}
					/>
				</tbody>
			</table>,
		);

		expect(asFragment()).toMatchSnapshot();
		expect(screen.getByTestId("Amount")).toHaveTextContent("10 LSK");
		expect(screen.getByTestId("UnlockableBalanceRow__status")).toHaveTextContent(
			isReady ? translations.TRANSACTION.UNLOCK_TOKENS.UNLOCKABLE : translations.TRANSACTION.UNLOCK_TOKENS.LOCKED,
		);

		// toggle checkbox

		expect(screen.getAllByRole("checkbox")).toHaveLength(1);

		userEvent.click(screen.getByRole("checkbox"));

		expect(onToggle).toHaveBeenCalledTimes(isReady ? 1 : 0);
	});
});
