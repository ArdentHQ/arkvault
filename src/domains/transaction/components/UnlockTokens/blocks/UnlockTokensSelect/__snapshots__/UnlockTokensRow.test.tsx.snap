// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnlockTokensRow > should render with isReady = false 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="group css-16ynu5g"
        data-testid="TableRow"
      >
        <td>
          <div
            class="flex px-3 items-center my-1 transition-colors duration-100 min-h-11 pl-3 ml-3 rounded-l dark:group-hover:bg-black group-hover:bg-theme-secondary-200"
          >
            <span
              class="whitespace-nowrap font-bold text-theme-secondary-900 dark:text-theme-secondary-200"
              data-testid="Amount"
            >
              10 LSK
            </span>
          </div>
        </td>
        <td>
          <div
            class="flex px-3 items-center my-1 transition-colors duration-100 min-h-11 dark:group-hover:bg-black group-hover:bg-theme-secondary-200"
          >
            <time
              class="text-theme-secondary-700"
              datetime="Wed, 01 Jan 2020 00:00:00 GMT"
            >
              01 Jan 2020 at 00:00:00 (6 months ago)
            </time>
          </div>
        </td>
        <td>
          <div
            class="flex px-3 items-center my-1 transition-colors duration-100 min-h-11 pr-3 mr-3 rounded-r dark:group-hover:bg-black group-hover:bg-theme-secondary-200 justify-end"
          >
            <div
              class="flex items-center"
            >
              <div
                class="flex items-center"
              >
                <span
                  class="text-theme-secondary-700"
                  data-testid="UnlockableBalanceRow__status"
                >
                  Locked
                </span>
                <div
                  class="ml-2 mr-3 text-theme-secondary-700"
                >
                  <div
                    style="height: 20px; width: 20px;"
                  >
                    <svg
                      id="lock"
                      style="height: 100%; width: 100%;"
                      viewBox="0 0 20 20"
                      x="0"
                      xml:space="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <g
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      >
                        <path
                          d="M3.4 9.3c0-.7.5-1.2 1.2-1.2h10.9c.7 0 1.2.5 1.2 1.2v8.5c0 .7-.5 1.2-1.2 1.2h-11c-.7 0-1.2-.6-1.2-1.2l.1-8.5h0zM5.8 8.1V5c.1-2.3 2.1-4.2 4.5-4 2.2.1 3.9 1.9 4 4v3M10 12.9h0"
                        />
                      </g>
                    </svg>
                  </div>
                </div>
              </div>
              <input
                class="custom-checkbox !h-5 !w-5 cursor-pointer !rounded !border-2 transition duration-150 ease-in-out focus:!ring-theme-primary-400 focus:!ring-offset-0 disabled:cursor-not-allowed disabled:border-theme-secondary-300 disabled:bg-theme-secondary-200 disabled:dark:border-theme-secondary-600 disabled:dark:bg-theme-secondary-800 [&:not(:checked)]:border-theme-secondary-300 [&:not(:checked)]:!bg-transparent [&:not(:checked)]:dark:border-theme-secondary-600 checked:text-theme-primary-600 checked:hover:text-theme-primary-700 [&:not(:checked)]:hover:border-theme-primary-600"
                disabled=""
                type="checkbox"
              />
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`UnlockTokensRow > should render with isReady = true 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="group css-16ynu5g"
        data-testid="TableRow"
      >
        <td>
          <div
            class="flex px-3 items-center my-1 transition-colors duration-100 min-h-11 pl-3 ml-3 rounded-l dark:group-hover:bg-black group-hover:bg-theme-secondary-200"
          >
            <span
              class="whitespace-nowrap font-bold text-theme-secondary-900 dark:text-theme-secondary-200"
              data-testid="Amount"
            >
              10 LSK
            </span>
          </div>
        </td>
        <td>
          <div
            class="flex px-3 items-center my-1 transition-colors duration-100 min-h-11 dark:group-hover:bg-black group-hover:bg-theme-secondary-200"
          >
            <time
              class="text-theme-secondary-700"
              datetime="Wed, 01 Jan 2020 00:00:00 GMT"
            >
              01 Jan 2020 at 00:00:00 (6 months ago)
            </time>
          </div>
        </td>
        <td>
          <div
            class="flex px-3 items-center my-1 transition-colors duration-100 min-h-11 pr-3 mr-3 rounded-r dark:group-hover:bg-black group-hover:bg-theme-secondary-200 justify-end"
          >
            <div
              class="flex items-center"
            >
              <div
                class="flex items-center"
              >
                <span
                  class="text-theme-secondary-700"
                  data-testid="UnlockableBalanceRow__status"
                >
                  Unlockable
                </span>
                <div
                  class="ml-2 mr-3 text-theme-primary-600"
                >
                  <div
                    style="height: 20px; width: 20px;"
                  >
                    <svg
                      style="height: 100%; width: 100%;"
                      viewBox="0 0 20 20"
                      x="0"
                      xml:space="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <path
                        d="M6.4 9.3c0-.7.5-1.2 1.1-1.2h10.3c.7 0 1.1.5 1.1 1.2v8.2c0 .7-.5 1.2-1.1 1.2H7.4c-.7 0-1.1-.6-1.1-1.2l.1-8.2z"
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                      <path
                        d="M13.1 11.8c.3 0 .5.2.5.5s-.2.5-.5.5-.5-.2-.5-.5c0-.2.2-.5.5-.5M1 8.2v-3c.1-2.2 2-4 4.3-3.8C7.3 1.5 9 3.2 9 5.2v2.9"
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <input
                checked=""
                class="custom-checkbox !h-5 !w-5 cursor-pointer !rounded !border-2 transition duration-150 ease-in-out focus:!ring-theme-primary-400 focus:!ring-offset-0 disabled:cursor-not-allowed disabled:border-theme-secondary-300 disabled:bg-theme-secondary-200 disabled:dark:border-theme-secondary-600 disabled:dark:bg-theme-secondary-800 [&:not(:checked)]:border-theme-secondary-300 [&:not(:checked)]:!bg-transparent [&:not(:checked)]:dark:border-theme-secondary-600 checked:text-theme-primary-600 checked:hover:text-theme-primary-700 [&:not(:checked)]:hover:border-theme-primary-600"
                type="checkbox"
              />
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
